src/dscanner/analysis/assert_without_msg.d

```diff
@@ -5,132 +5,165 @@
 module dscanner.analysis.assert_without_msg;
 
 import dscanner.analysis.base;
-import dscanner.utils : safeAccess;
-import dsymbol.scope_ : Scope;
-import dparse.lexer;
-import dparse.ast;
-
 import std.stdio;
-import std.algorithm;
 
 /**
  * Check that all asserts have an explanatory message.
  */
-final class AssertWithoutMessageCheck : BaseAnalyzer
+extern(C++) class AssertWithoutMessageCheck(AST) : BaseAnalyzerDmd!AST
 {
-	enum string KEY = "dscanner.style.assert_without_msg";
-	enum string MESSAGE = "An assert should have an explanatory message";
 	mixin AnalyzerInfo!"assert_without_msg";
+	alias visit = BaseAnalyzerDmd!AST.visit;
 
 	///
-	this(string fileName, const(Scope)* sc, bool skipTests = false)
+	extern(D) this(string fileName, bool skipTests = false)
 	{
-		super(fileName, sc, skipTests);
+		super(fileName, skipTests);
+	
+		import std.file : readText;
+		this.code = readText(fileName);
 	}
 
-	override void visit(const AssertExpression expr)
+	mixin ScopedVisit!(AST.FuncDeclaration);
+	mixin ScopedVisit!(AST.IfStatement);
+	mixin ScopedVisit!(AST.WhileStatement);
+	mixin ScopedVisit!(AST.ForStatement);
+	mixin ScopedVisit!(AST.ForeachStatement);
+	mixin ScopedVisit!(AST.ClassDeclaration);
+	mixin ScopedVisit!(AST.StructDeclaration);
+	mixin ScopedVisit!(AST.InterfaceDeclaration);
+	mixin ScopedVisit!(AST.UnionDeclaration);
+	mixin ScopedVisit!(AST.ScopeStatement);
+
+	override void visit(AST.AssertExp ae)
 	{
-		if (expr.assertArguments && expr.assertArguments.message is null)
-			addErrorMessage(expr.line, expr.column, KEY, MESSAGE);
+		// We need to ensure that this is an actual `assert` expression,
+		// Not an in/out contract represented also as an AssertExp in the AST
+		import std.string : indexOf;
+
+		// File Offset indicates where the expression starts
+		// To extract the function name get rid of `(` and everything that follows
+		string funcName = code[ae.loc.fileOffset .. $];
+		funcName = funcName[0 .. funcName.indexOf("(")];
+
+		if (funcName == "assert" && !ae.msg)
+		// if (!ae.msg)
+			addErrorMessage(ae.loc.linnum, ae.loc.charnum, KEY, MESSAGE);
 	}
 
-	override void visit(const FunctionCallExpression expr)
+	override void visit(AST.StaticAssert ae)
 	{
-		if (!isStdExceptionImported)
+		if (!ae.msg)
+			addErrorMessage(ae.loc.linnum, ae.loc.charnum, KEY, MESSAGE);
+	}
+
+	override void visit(AST.Import i)
+	{
+		import std.algorithm.iteration : map;
+		import std.array : array;
+
+		const(char[])[] exceptionImportStr = ["std", "exception"];
+		
+		const(char[])[] aux = i.packages.map!(a => a.toString()).array;
+		aux ~= i.id.toString();
+
+		// import std.exception without selective imports
```

Comment:
This is a brittle check. `import std.exception` could be at any scope. you assume that the import is global.

Reply:
I am keeping track of the local imports, we actually have an `unittest` for that.
```
unittest {
		import exception;
		class C {
			import std.exception;
		}
		enforce(0); // std.exception not imported yet - could be a user-defined symbol
		struct S {
			import std.exception;
		}
		enforce(0); // std.exception not imported yet - could be a user-defined symbol
		if (false) {
			import std.exception;
		}
		enforce(0); // std.exception not imported yet - could be a user-defined symbol
		{
			import std.exception;
		}
		enforce(0); // std.exception not imported yet - could be a user-defined symbol
	}
```

Reply:
Ok.